src/layout.c

```diff
@@ -1115,21 +1116,158 @@ int igraph_layout_reingold_tilford(const igraph_t *graph,
   }
   
   /* ----------------------------------------------------------------------- */
-  /* Ok, we have the root vertices now. 
-     Possibly we need to add some edges to make the graph connected. */
+  /* Ok, we have the root vertices. 
+
+     Now, the function that computes the layout is internal only (no API):
+     igraph_i_layout_reingold_tilford. That function needs a connected graph
+     since only then you can compute the coordinates, but we want to extend
+     the algorithm to other cases, including:
+     - undirected disconnected graphs
+     - weakly but not strongly connected directed graphs
+
+     An extreme case a directed tree that goes in the opposite direction
+     than the mode selected for the RT layout, e.g. an "out"-directed
+     tree that is asked to draw an RT layout with mode "in". In this case,
+     the graph is weakly connected but its strong connectivity is all
+     upside down. (No matter that such a layout won't be very useful
+     in practice.)
+
+     To deal with these extensions of the original algorithm, we construct
+     an "extended graph" with "fake" edges that ensure full connectivity
+     for an undirected graph or full strong connectivity for a directed graph.
+     This extended graph is then passed to the layout factory.
+
+     For undirected graphs, we just connect the roots to a single, top
+     "real_root" node.
+
+     For directed graphs, it's more complicated. Above we computed one root
+     per weakly connected component, but there might be nodes within that
+     component that cannot be reached from that root in a directed sense, e.g.:
 
+     1
+     |
+     v
+     2 <- 3
+
+     If the root is 1, how do you reach 3?
+
+     Although more clever ideas are envisioned for future releases, for now
+     just make a list of all nodes of this kind and add directed edges
+     pointing directly from the root to them.
+
+     Then, just like for undirected graphs, add a fake "real_root" and
+     put all roots as directed children of it.
+
+     */
+
+  /* the graph is connected, good */
   if (igraph_vector_size(proots)==1) {
     real_root=(long int) VECTOR(*proots)[0];
     if (real_root<0 || real_root>=no_of_nodes) {
       IGRAPH_ERROR("invalid vertex id", IGRAPH_EINVVID);
     }
-  } else {
+  }
+
+  /* the graph is disconnected and directed, add edges within
+     weakly connected components from the root to all nodes that
+     cannot be reached
+     
+     TODO: a better strategy would be nice here
+     */
+  if (igraph_vector_size(proots) != 1 && mode != IGRAPH_ALL) {
+
+    igraph_vector_t newedges;
+    long int no_of_newedges, root;
```

Comment:
`no_of_newedges` is never initialized, and should be set to 0 at some point. This yields a (possible) error later on with memory allocation when initializing the vector.